app/Components/Sudoku Board/SudokuBoard.tsx

```diff
@@ -434,11 +434,30 @@ const SudokuBoard = (props: SudokuBoardProps) => {
   };
 
   const handleKeyDown = (event: any) => {
+    const inputValue = event.nativeEvent.key;
+
+    if (inputValue == "u" || inputValue == "U") {
+      const isUndoButtonDisabled =
+        sudokuBoard.actionHistory == null ||
+        sudokuBoard.actionHistory.length == 0;
+      if (!isUndoButtonDisabled) {
+        undo();
+      }
+    } else if (inputValue == "p" || inputValue == "P") {
+      handlePause(sudokuBoard, navigation);
+    } else if (
+      inputValue == "t" ||
+      inputValue == "T" ||
+      inputValue == "n" ||
+      inputValue == "N"
+    ) {
+      toggleNoteMode();
+    }
+
     if (sudokuBoard.selectedCell == null) {
       return;
     }
 
-    const inputValue = event.nativeEvent.key;
     if (/^[1-9]$/.test(inputValue)) {
       updateCellEntry(parseInt(inputValue, 10));
     } else if (
```

Comment:
nit: this else if can be added to above if (or switch) statement

Reply:
I moved the else if to bottom switch statement. I kept the regex outside of the switch statement as I wasn't sure how to get it working inside, aside from doing:
```javascript
case "1":
case "2":
```
etc and I didn't want to do that.